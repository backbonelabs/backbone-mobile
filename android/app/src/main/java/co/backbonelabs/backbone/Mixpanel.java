package co.backbonelabs.backbone;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.mixpanel.android.mpmetrics.MixpanelAPI;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * React Native wrapper for the Mixpanel Android API
 * Most of this code was copied from https://github.com/davodesign84/react-native-mixpanel
 */
public class Mixpanel extends ReactContextBaseJavaModule implements LifecycleEventListener {
    private MixpanelAPI mixpanel;

    public Mixpanel(ReactApplicationContext reactContext) {
        super(reactContext);
        String projectToken = BuildConfig.MIXPANEL_TOKEN;
        mixpanel = MixpanelAPI.getInstance(reactContext.getCurrentActivity(), projectToken);
    }

    /**
     * Converts a ReadableMap object from React Native to a JSONObject.
     * @param readableMap The ReadableMap object to convert to a JSONObject
     * @return JSONObject equivalent of readableMap
     * @throws JSONException
     */
    static JSONObject reactToJSON(ReadableMap readableMap) throws JSONException {
        JSONObject jsonObject = new JSONObject();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
        while(iterator.hasNextKey()){
            String key = iterator.nextKey();
            ReadableType valueType = readableMap.getType(key);
            switch (valueType){
                case Null:
                    jsonObject.put(key,JSONObject.NULL);
                    break;
                case Boolean:
                    jsonObject.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    jsonObject.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    jsonObject.put(key, readableMap.getString(key));
                    break;
                case Map:
                    jsonObject.put(key, reactToJSON(readableMap.getMap(key)));
                    break;
                case Array:
                    jsonObject.put(key, reactToJSON(readableMap.getArray(key)));
                    break;
            }
        }

        return jsonObject;
    }

    /**
     * Converts a ReadableArray object from React Native to a JSONObject.
     * @param readableArray The ReadableArray object to convert to a JSONObject
     * @return JSONObject equivalent of readableArray
     * @throws JSONException
     */
    static JSONArray reactToJSON(ReadableArray readableArray) throws JSONException {
        JSONArray jsonArray = new JSONArray();
        for(int i=0; i < readableArray.size(); i++) {
            ReadableType valueType = readableArray.getType(i);
            switch (valueType){
                case Null:
                    jsonArray.put(JSONObject.NULL);
                    break;
                case Boolean:
                    jsonArray.put(readableArray.getBoolean(i));
                    break;
                case Number:
                    jsonArray.put(readableArray.getDouble(i));
                    break;
                case String:
                    jsonArray.put(readableArray.getString(i));
                    break;
                case Map:
                    jsonArray.put(reactToJSON(readableArray.getMap(i)));
                    break;
                case Array:
                    jsonArray.put(reactToJSON(readableArray.getArray(i)));
                    break;
            }
        }
        return jsonArray;
    }

    @Override
    public String getName() {
        return "Mixpanel";
    }

    /**
     * Track an event with no properties
     * @param name The name of the event to send
     */
    @ReactMethod
    public void track(final String name) {
        mixpanel.track(name);
    }

    /**
     * Track an event with properties
     * @param name The name of the event to send
     * @param props Key-value pairs of the properties to include in the event
     */
    @ReactMethod
    public void trackWithProperties(final String name, final ReadableMap props) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(props);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.track(name, obj);
    }

    /**
     * Create an alias for linking a user ID to the current ID
     * @param newId New ID to link to current ID
     */
    @ReactMethod
    public void createAlias(final String newId) {
        mixpanel.alias(newId, mixpanel.getDistinctId());
    }

    /**
     * Associate future calls to track(String) with the user identified by the given distinct ID
     * @param userId A unique identifier for the user
     */
    @ReactMethod
    public void identify(final String userId) {
        mixpanel.identify(userId);
        mixpanel.getPeople().identify(userId);
    }

    /**
     * Begin timing of an event
     * @param event The name of the event to track with timing
     */
    @ReactMethod
    public void timeEvent(final String event) {
        mixpanel.timeEvent(event);
    }

    /**
     * Register properties that will be sent with every subsequent call to track(String)
     * @param properties Super properties to register
     */
    @ReactMethod
    public void registerSuperProperties(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.registerSuperProperties(obj);
    }

    /**
     * Register super properties like registerSuperProperties(ReadableMap), but only
     * if no other super property with the same names have already been registered
     * @param properties Super properties to register
     */
    @ReactMethod
    public void registerSuperPropertiesOnce(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.registerSuperPropertiesOnce(obj);
    }

    /**
     * Set a collection of properties on a user all at once
     * @param properties Key-value pair of property names and values
     */
    @ReactMethod
    public void set(final ReadableMap properties) {
        JSONObject obj = null;
        try {
            obj = Mixpanel.reactToJSON(properties);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        mixpanel.getPeople().set(obj);
    }

    /**
     * Increment an existing numbered user property
     * @param property The user property to update
     * @param count The amount to be added to the current value of the property
     */
    @ReactMethod
    public void incrementUserPropertyBy(final String property, final double count) {
        mixpanel.getPeople().increment(property, count);
    }

    /**
     * Clear all distinct IDs, super properties, and push registrations from persistent storage,
     * and then flush the events
     */
    @ReactMethod
    public void reset() {
        mixpanel.reset();
        mixpanel.flush();
    }

    /**
     * Push all queued events and user profile changes to Mixpanel
     */
    @ReactMethod
    public void flush() {
        mixpanel.flush();
    }

    @Override
    public void onHostResume() {
        // Actvity `onResume`
    }

    @Override
    public void onHostPause() {
        // Actvity `onPause`

        if (mixpanel != null) {
            mixpanel.flush();
        }
    }

    @Override
    public void onHostDestroy() {
        // Actvity `onDestroy`

        if (mixpanel != null) {
            mixpanel.flush();
        }
    }
}
