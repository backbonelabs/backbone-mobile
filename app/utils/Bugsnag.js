import { Client } from 'bugsnag-react-native';
import { NativeModules } from 'react-native';

const bugsnag = new Client(NativeModules.Environment.BUGSNAG_API_KEY);

// See https://docs.bugsnag.com/platforms/react-native/ for Bugsnag client API

export default bugsnag;
