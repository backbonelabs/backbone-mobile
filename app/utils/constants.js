export default {
  storageNamespace: 'backbone',
  storageKeys: {
    ACCESS_TOKEN: 'accessToken',
    USER: 'user',
    DEVICE: 'device',
    CALIBRATION_AUTO_START: 'calibrationAutoStart',
    SESSION_STATE: 'sessionParameters',
    APP_RATING_THRESHOLD: 'appRatingThreshold',
  },
  bluetoothStates: {
    UNKNOWN: -1,
    UNSUPPORTED: 0,
    RESETTING: 1,
    OFF: 2,
    TURNING_ON: 3,
    ON: 4,
    DISCONNECTED: 5,
    CONNECTING: 6,
    CONNECTED: 7,
    DISCONNECTING: 8,
    TURNING_OFF: 9,
  },
  deviceModes: {
    UNKNOWN: 0,
    BACKBONE: 1,
    BOOTLOADER: 2,
  },
  deviceStatuses: {
    DISCONNECTED: 0,
    CONNECTING: 1,
    CONNECTED: 2,
    DISCONNECTING: 3,
  },
  sessionOperations: {
    START: 0,
    RESUME: 1,
    PAUSE: 2,
    STOP: 3,
  },
  notificationTypes: {
    SLOUCH_WARNING: 100,
    FOREGROUND_SERVICE: 101,
    INACTIVITY_REMINDER: 102,
    DAILY_REMINDER: 103,
    SINGLE_REMINDER: 104,
    INFREQUENT_REMINDER: 105,
    SESSION_COMPLETED: 106,
  },
  vibrationDurations: {
    SHORT: 200,
    MEDIUM: 500,
    LONG: 800,
  },
  firmwareUpdateStates: {
    INVALID_SERVICE: -2,
    INVALID_FILE: -1,
    BEGIN: 0,
    END_SUCCESS: 1,
    END_ERROR: 2,
  },
  firmwareUpdateErrorCodes: {
    COMMAND_RESULT: 300,
    COMMAND_VERIFY: 301,
    UPDATE_VALUE: 302,
    WRITE_VALUE: 303,
    ROW_NUMBER: 304,
  },
  firmwareUpdateCommandCodes: {
    VERIFY_CHECKSUM: 49,
    GET_FLASH_SIZE: 50,
    SEND_DATA: 55,
    ENTER_BOOTLOADER: 56,
    PROGRAM_ROW: 57,
    VERIFY_ROW: 58,
    EXIT_BOOTLOADER: 59,
  },
  errorMessages: {
    NETWORK_ERROR: 'Unexpected error. Please try again later.',
  },
  heightUnitIdToLabel: {
    1: 'IN',
    2: 'CM',
  },
  weightUnitIdToLabel: {
    1: 'LB',
    2: 'KG',
  },
  gender: {
    male: 1,
    female: 2,
    other: 3,
  },
  height: {
    defaults: {
      value: 60,
      unit: 1,
    },
    units: {
      IN: 1,
      CM: 2,
    },
    conversionValue: 2.54,
  },
  weight: {
    defaults: {
      value: 100,
      unit: 1,
    },
    units: {
      LB: 1,
      KG: 2,
    },
    conversionValue: 0.453592,
  },
  months: [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ],
  /* eslint-disable max-len */
  emailRegex: /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i,
  /* eslint-disable max-len */
  surveyUrls: {
    baseline: 'https://backbonelabsinc.typeform.com/to/lVs1Sh',
    feedback: 'https://backbonelabsinc.typeform.com/to/rmq85N',
  },
  surveyThresholds: {
    feedbackSurvey: 3,
    initialAppRating: 5,
    additionalAppRating: 4,
  },
  appUrls: {
    ios: 'https://appsto.re/us/1vAOgb.i',
    android: 'https://play.google.com/store/apps/details?id=co.backbonelabs.backbone',
  },
  authMethods: {
    EMAIL: 1,
    FACEBOOK: 2,
  },
};
