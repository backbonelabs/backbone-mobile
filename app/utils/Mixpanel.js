import { NativeModules } from 'react-native';
import { map } from 'lodash';
import constants from './constants';

const { Mixpanel, BluetoothService } = NativeModules;

export default {
  /**
   * Associates all future events with the specified user id
   * @param  {String}  userId  User account id
   */
  identify(userId) {
    Mixpanel.identify(userId);
  },

  /**
   * Sets the user's profile properties on Mixpanel
   * @param  {Object}  userDetails  User profile properties
   */
  set(userDetails) {
    Mixpanel.set(userDetails);
  },

  /**
   * Formats the user profile properties before setting on Mixpanel
   * @param  {Object}  userDetails  User profile properties
   */
  setUserProperties(user) {
    const userClone = Object.assign({ ...user }, { ...user.settings });
    const { gender, heightUnitPreference, weightUnitPreference } = user;
    const { height, weight } = constants;
    const formattedGender = gender === constants.gender.male ? 'male' : 'female';

    // Delete unwanted profile properties
    [
      '_id',
      'confirmationToken',
      'confirmationTokenExpiry',
      'firstName',
      'lastName',
      'nickname',
      'settings',
    ].forEach(value => {
      delete userClone[value];
    });


    map(userClone, (value, key) => {
      if (value || value === false) {
        if (key === 'email') {
          // Set Mixpanel special property $email and delete unused email property
          userClone.$email = user[key];
          delete userClone[key];
        } else if (value && key === 'gender') {
          // Change gender from integer to word equivalent ("male" / "female")
          userClone.gender = formattedGender;
        } else if (key === 'height') {
          // Change heightUnitPreference from integer to word equivalent ("IN" / "CM")
          userClone.heightUnitPreference = heightUnitPreference === height.units.IN ? 'IN' : 'CM';
        } else if (key === 'weight') {
          // Change heightUnitPreference from integer to word equivalent ("LB" / "KG")
          userClone.weightUnitPreference = weightUnitPreference === weight.units.LB ? 'LB' : 'KG';
        }
      } else {
        if (key === 'weight' || key === 'height') {
          // These unit preference properties will never be undefined
          // Remove them if their measurement values are undefined
          delete userClone[`${key}UnitPreference`];
        }

        // Remove properties with undefined values
        delete userClone[key];
      }
    });

    this.set(userClone);
  },

  /**
   * Tracks an event
   * @param  {String}  event  Event name
   */
  track(event) {
    Mixpanel.track(event);
  },

  /**
   * Tracks an event with properties
   * @param  {String}  event       Event name
   * @param  {Object}  properties  An object containing properties relevant to the event
   */
  trackWithProperties(event, properties) {
    Mixpanel.trackWithProperties(event, properties);
  },

  /**
   * Tracks an error event and its properties
   * @param  {Object}  properties  An error event property object which contains the error message,
   *                               stack trace, file path, bluetooth state properties, and anything
   *                               else that you feel would be relevant to the error event.
   */
  trackError(properties) {
    // Track Bluetooth status to identify potential BT-related issues
    BluetoothService.getState((error, { state }) => {
      Mixpanel.trackWithProperties('reactNativeError', {
        bluetoothState: state,
        ...properties,
      });
    });
  },

  /**
   * Registers super properties, which Mixpanel will include with every event
   * @param  {Object}  properties  An object containing properties to pass on every event
   */
  registerSuperProperties(properties) {
    Mixpanel.registerSuperProperties(properties);
  },
};
