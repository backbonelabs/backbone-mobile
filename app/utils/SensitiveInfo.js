import SInfo from 'react-native-sensitive-info';
import constants from './constants';

const { storageNamespace } = constants;

const options = {
  sharedPreferencesName: storageNamespace, // Android
  keychainService: storageNamespace, // iOS
};

/**
 * Retrieves a value at a given key. The resolved value will be parsed as JSON.
 * @param  {String}  key The key the value is stored in
 * @return {Promise}
 */
const getItem = key => SInfo.getItem(key, options)
  .then(value => {
    try {
      return JSON.parse(value);
    } catch (err) {
      return value;
    }
  });

/**
 * Sets a value at a given key. If the value is a string, it will be stored
 * as a string. If the value is any other type, it will be stringified using
 * JSON.stringify before being stored. If value is undefined, it is a no-op.
 * @param {String} key   The key to store value in
 * @param {*}      value The value to store
 */
const setItem = (key, value) => {
  if (value === undefined) {
    return;
  }
  let _value = value;
  if (typeof _value !== 'string') {
    _value = JSON.stringify(_value);
  }
  SInfo.setItem(key, _value, options);
};

/**
 * Removes a key-value pair.
 * @param {String} key The key the value is stored in
 */
const deleteItem = key => SInfo.deleteItem(key, options);

export default { getItem, setItem, deleteItem };
