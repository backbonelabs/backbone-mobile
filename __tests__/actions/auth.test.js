import configureStore from 'redux-mock-store';
import { NativeModules } from 'react-native';
import thunk from 'redux-thunk';
import { asyncActionMiddleware } from 'redux-async-action';
import authActions from '../../app/actions/auth';
import Bugsnag from '../../app/utils/Bugsnag';
import Mixpanel from '../../app/utils/Mixpanel';
import SensitiveInfo from '../../app/utils/SensitiveInfo';
import constants from '../../app/utils/constants';

const { storageKeys } = constants;
const { UserService } = NativeModules;

describe('__Auth Actions__', () => {
  const store = configureStore([asyncActionMiddleware, thunk])({});

  const user = {
    email: 'test@mail.com',
    nickname: 'testNick',
    password: 'password',
    _id: 'testId',
    accessToken: 'testToken',
  };
  const { email, password } = user;

  beforeEach(() => {
    store.clearActions();
    UserService.setUserId.mockClear();
    Bugsnag.setUser.mockClear();
    Mixpanel.identify.mockClear();
    Mixpanel.setUserProperties.mockClear();
    Mixpanel.track.mockClear();
    Mixpanel.trackWithProperties.mockClear();
    SensitiveInfo.setItem.mockClear();
  });

  test('creates an action to Login', async () => {
    const { accessToken, ...userObj } = user;
    fetch.mockResponseSuccess(user);
    await store.dispatch(authActions.login({ email, password }));

    expect(UserService.setUserId).toBeCalledWith(user._id);
    expect(Bugsnag.setUser).toBeCalledWith(user._id, user.nickname, user.email);
    expect(Mixpanel.identify).toBeCalledWith(user._id);
    expect(Mixpanel.setUserProperties).toBeCalledWith(userObj);
    expect(Mixpanel.track).toHaveBeenCalledWith('login-success');
    expect(SensitiveInfo.setItem.mock.calls[0]).toEqual([storageKeys.ACCESS_TOKEN, accessToken]);
    expect(SensitiveInfo.setItem.mock.calls[1]).toEqual([storageKeys.USER, userObj]);
    expect(fetch.mock.calls[0][1].body).toEqual(JSON.stringify({ email, password }));
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when Login is unsuccessful', async () => {
    fetch.mockResponseSuccess({ error: 'api server error' });
    await store.dispatch(authActions.login());
    expect(Mixpanel.trackWithProperties)
    .toBeCalledWith('login-error', { errorMessage: 'api server error' });
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when Login returns a server error', async () => {
    fetch.mockResponseFailure();
    await store.dispatch(authActions.login());
    expect(Mixpanel.track).toHaveBeenCalledWith('login-serverError');
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action to Signup', async () => {
    const { accessToken, ...userObj } = user;
    const body = { user: { ...userObj } };
    fetch.mockResponseSuccess({ ...body, accessToken });
    await store.dispatch(authActions.signup({ email, password }));
    expect(UserService.setUserId).toBeCalledWith(body.user._id);
    expect(Bugsnag.setUser).toBeCalledWith(body.user._id, body.user.nickname, body.user.email);
    expect(Mixpanel.identify).toBeCalledWith(body.user._id);
    expect(Mixpanel.setUserProperties).toBeCalledWith(body.user);
    expect(Mixpanel.track).toHaveBeenCalledWith('signup-success');
    expect(SensitiveInfo.setItem.mock.calls[0]).toEqual([storageKeys.ACCESS_TOKEN, accessToken]);
    expect(SensitiveInfo.setItem.mock.calls[1]).toEqual([storageKeys.USER, body.user]);
    expect(fetch.mock.calls[0][1].body).toEqual(JSON.stringify({ email, password }));
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when Signup is unsuccessful', async () => {
    fetch.mockResponseSuccess({ error: 'api server error' });
    await store.dispatch(authActions.signup());
    expect(Mixpanel.trackWithProperties)
    .toBeCalledWith('signup-error', { errorMessage: 'api server error' });
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when Signup returns a server error', async () => {
    fetch.mockResponseFailure();
    await store.dispatch(authActions.signup());
    expect(Mixpanel.track).toHaveBeenCalledWith('signup-serverError');
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action to reset password', async () => {
    fetch.mockResponseSuccess({ ok: true });
    await store.dispatch(authActions.reset({ email: 'test@mail.com' }));
    expect(Mixpanel.trackWithProperties)
    .toBeCalledWith('passwordReset-success', { email: 'test@mail.com' });
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when resetting password is unsuccessful', async () => {
    fetch.mockResponseSuccess({
      error: 'api server error',
    });
    await store.dispatch(authActions.reset({ email: 'test@mail.com' }));
    expect(Mixpanel.trackWithProperties)
    .toBeCalledWith(
      'passwordReset-error',
      { email: 'test@mail.com', errorMessage: 'api server error' },
    );
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates an action when resetting password returns a server error', async () => {
    fetch.mockResponseFailure();
    await store.dispatch(authActions.reset({ email: 'test@mail.com' }));
    expect(Mixpanel.track).toHaveBeenCalledWith('passwordReset-serverError');
    expect(store.getActions()).toMatchSnapshot();
  });

  test('creates a SIGN_OUT action', () => {
    store.dispatch(authActions.signOut());
    expect(UserService.unsetUserId).toHaveBeenCalled();
    expect(Bugsnag.clearUser).toHaveBeenCalled();
    expect(Mixpanel.track).toHaveBeenCalledWith('signOut');
    expect(SensitiveInfo.deleteItem.mock.calls[0]).toEqual([storageKeys.ACCESS_TOKEN]);
    expect(SensitiveInfo.deleteItem.mock.calls[1]).toEqual([storageKeys.USER]);
    expect(authActions.signOut()).toMatchSnapshot();
  });

  test('creates a SET_ACCESS_TOKEN action', () => {
    expect(authActions.setAccessToken('testToken')).toMatchSnapshot();
  });
});
