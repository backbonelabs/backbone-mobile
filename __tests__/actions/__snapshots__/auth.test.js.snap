// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Actions creates a SET_ACCESS_TOKEN action 1`] = `
Object {
  "payload": "testToken",
  "type": "SET_ACCESS_TOKEN",
}
`;

exports[`Auth Actions creates a SIGN_OUT action 1`] = `
Object {
  "type": "SIGN_OUT",
}
`;

exports[`Auth Actions creates an action to Login 1`] = `
Array [
  Object {
    "type": "LOGIN__START",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "_id": "testId",
      "accessToken": "testToken",
      "authMethod": 1,
      "email": "test@mail.com",
      "nickname": "testNick",
      "password": "password",
    },
    "type": "LOGIN",
  },
]
`;

exports[`Auth Actions creates an action to Signup 1`] = `
Array [
  Object {
    "type": "SIGNUP__START",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "accessToken": "testToken",
      "user": Object {
        "_id": "testId",
        "authMethod": 1,
        "email": "test@mail.com",
        "nickname": "testNick",
        "password": "password",
      },
    },
    "type": "SIGNUP",
  },
]
`;

exports[`Auth Actions creates an action to reset password 1`] = `
Array [
  Object {
    "type": "PASSWORD_RESET__START",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "PASSWORD_RESET",
  },
]
`;

exports[`Auth Actions creates an action when Login is unsuccessful 1`] = `
Array [
  Object {
    "type": "LOGIN__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: api server error],
    "type": "LOGIN__ERROR",
  },
]
`;

exports[`Auth Actions creates an action when Login returns a server error 1`] = `
Array [
  Object {
    "type": "LOGIN__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: Unexpected error. Please try again later.],
    "type": "LOGIN__ERROR",
  },
]
`;

exports[`Auth Actions creates an action when Signup is unsuccessful 1`] = `
Array [
  Object {
    "type": "SIGNUP__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: api server error],
    "type": "SIGNUP__ERROR",
  },
]
`;

exports[`Auth Actions creates an action when Signup returns a server error 1`] = `
Array [
  Object {
    "type": "SIGNUP__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: Unexpected error. Please try again later.],
    "type": "SIGNUP__ERROR",
  },
]
`;

exports[`Auth Actions creates an action when resetting password is unsuccessful 1`] = `
Array [
  Object {
    "type": "PASSWORD_RESET__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: api server error],
    "type": "PASSWORD_RESET__ERROR",
  },
]
`;

exports[`Auth Actions creates an action when resetting password returns a server error 1`] = `
Array [
  Object {
    "type": "PASSWORD_RESET__START",
  },
  Object {
    "error": true,
    "meta": undefined,
    "payload": [Error: Unexpected error. Please try again later.],
    "type": "PASSWORD_RESET__ERROR",
  },
]
`;
